### PR TITLE
fix shadow and EDL not being exported with 3D animation tool

### PR DESCRIPTION
**Pull request by NEDJIMAbelgacem**
## Description
Fixes the issue submitted here: https://github.com/qgis/QGIS/issues/41648

A sample of exported animation of point cloud with shadows enabled:
https://user-images.githubusercontent.com/29183781/108865375-7aa3f400-75f3-11eb-9ca0-09a6e35fc3eb.mp4

### CODE REVIEW

File: src/3d/qgsoffscreen3dengine.cpp [hunk 1]
```diff
@@ -36,29 +36,8 @@
...
-  {
-    format.setVersion( 4, 3 );
-    format.setProfile( QSurfaceFormat::CoreProfile );
-  }
-#endif
-
-  format.setMajorVersion( 3 );
-  format.setDepthBufferSize( 32 ); // TODO: or 24?  (used by QWindow3D)
-  format.setSamples( 8 );
-  QSurfaceFormat::setDefaultFormat( format );
```
Comment (wonder-sk) [inline src/3d/qgsoffscreen3dengine.cpp:61]: Why is this removed - was it causing any problems?

File: src/3d/qgsoffscreen3dengine.cpp [hunk 2]
```diff
@@ -238,17 +143,11 @@ QSurface *QgsOffscreen3DEngine::surface() const
...
+  Qt3DRender::QRenderCaptureReply *captureReply;
+  captureReply = mFrameGraph->renderCapture()->requestCapture();
+  connect( captureReply, &Qt3DRender::QRenderCaptureReply::completed, this, [ = ]
   {
-    QImage image = mReply->image();
-    mReply->deleteLater();
-    mReply = nullptr;
-    emit imageCaptured( image );
+    emit imageCaptured( captureReply->image() );
+    captureReply->deleteLater();
```
Comment (wonder-sk) [inline src/3d/qgsoffscreen3dengine.cpp:151]: Just wondering, does this change fix anything regarding capture of images, or it is only changed to remove `mReply` from member variables?

File: src/3d/qgsshadowrenderingframegraph.cpp [hunk 3]
```diff
@@ -81,8 +79,10 @@ Qt3DRender::QFrameGraphNode *QgsShadowRenderingFrameGraph::constructForwardRende
   mForwardClearBuffers->setClearColor( QColor::fromRgbF( 0.0, 1.0, 0.0, 1.0 ) );
   mForwardClearBuffers->setBuffers( Qt3DRender::QClearBuffers::ColorDepthBuffer );
 
-  mFrustumCulling = new Qt3DRender::QFrustumCulling;
-  mFrustumCulling->setParent( mForwardClearBuffers );
+  mFrustumCulling = new Qt3DRender::QFrustumCulling( mForwardClearBuffers );
+
+  // mMainCameraSelector = new Qt3DRender::QCameraSelector( mForwardRenderTargetSelector );
+  // mMainCameraSelector->setCamera( mMainCamera );
```
Comment (wonder-sk) [inline src/3d/qgsshadowrenderingframegraph.cpp:85]: remove commented code if not needed?

File: src/3d/qgsshadowrenderingframegraph.cpp [hunk 4]
```diff
@@ -133,11 +133,54 @@ Qt3DRender::QFrameGraphNode *QgsShadowRenderingFrameGraph::constructPostprocessi
   mPostprocessClearBuffers = new Qt3DRender::QClearBuffers( mPostprocessPassLayerFilter );
   mPostprocessClearBuffers->setClearColor( QColor::fromRgbF( 0.0f, 0.0f, 0.0f ) );
 
+  //
+  {
```
Comment (wonder-sk) [inline src/3d/qgsshadowrenderingframegraph.cpp:137]: do we need the code to be inside the block?

File: src/3d/qgsshadowrenderingframegraph.cpp [hunk 5]
```diff
@@ -133,11 +133,54 @@ Qt3DRender::QFrameGraphNode *QgsShadowRenderingFrameGraph::constructPostprocessi
...
+    mRenderCaptureDepthTexture->setComparisonMode( Qt3DRender::QAbstractTexture::CompareRefToTexture );
+
+    depthOutput->setTexture( mRenderCaptureDepthTexture );
+    renderTarget->addOutput( depthOutput );
+
+    mRenderCaptureTargetSelector->setTarget( renderTarget );
+  }
+  //
+
+  mRenderCapture = new Qt3DRender::QRenderCapture( mRenderCaptureTargetSelector );
```
Comment (wonder-sk) [inline src/3d/qgsshadowrenderingframegraph.cpp:178]: isn't the render capture and its textures eat some of memory resources even if we don't need to capture frames? Or is that negligible? (I am wondering if it is not better to attach the textures and render capture only when we need them - but if that adds complexity, probably not worth it)

File: src/3d/qgsoffscreen3dengine.cpp (hunk 1 above)
Comment (NEDJIMAbelgacem) [reply src/3d/qgsoffscreen3dengine.cpp:61]: I thought it might be better to just have default surface format stuff, I don't see the need to change it

File: src/3d/qgsoffscreen3dengine.cpp (hunk 2 above)
Comment (NEDJIMAbelgacem) [reply src/3d/qgsoffscreen3dengine.cpp:151]: It is changed to make things the same between QgsWindow3DEngine and QgsOffscreen3DEngine, should I move it up to QgsAbstract3DEngine ?

File: src/3d/qgsshadowrenderingframegraph.cpp (hunk 4 above)
Comment (NEDJIMAbelgacem) [reply src/3d/qgsshadowrenderingframegraph.cpp:137]: Not really needed, it is just to make sure that I only pick up mRenderCaptureTargetSelector from that context (previously I was using the same prefix in the objects names).

File: src/3d/qgsshadowrenderingframegraph.cpp (hunk 5 above)
Comment (NEDJIMAbelgacem) [reply src/3d/qgsshadowrenderingframegraph.cpp:178]: I think you are right, I will try that and compare the frame rate

File: src/3d/qgsoffscreen3dengine.cpp (hunk 1 above)
Comment (wonder-sk) [reply src/3d/qgsoffscreen3dengine.cpp:61]: There is a comment which explains the reasoning:
```
  // by default we get just some older version of OpenGL from the system,
  // but for 3D lines we use "primitive restart" functionality supported in OpenGL >= 3.1
  // Qt3DWindow uses this - requesting OpenGL 4.3 - so let's request the same version.
```
Can you please check what is the OpenGL version you are getting and what is the default format you are getting?

File: src/3d/qgsoffscreen3dengine.cpp (hunk 2 above)
Comment (wonder-sk) [reply src/3d/qgsoffscreen3dengine.cpp:150]: Hmm I see... If both engines support capturing of renders, then maybe it's best to have the implementation in the parent class (especially if the implementations are the same)

File: src/3d/qgsoffscreen3dengine.cpp (hunk 1 above)
Comment (NEDJIMAbelgacem) [reply src/3d/qgsoffscreen3dengine.cpp]: After I turned back this bit, more tests were passing so I think I should just keep it as.

File: src/3d/qgsoffscreen3dengine.cpp (hunk 2 above)
Comment (NEDJIMAbelgacem) [reply src/3d/qgsoffscreen3dengine.cpp]: The implementation is the same and it is moved up now.

File: src/3d/qgsshadowrenderingframegraph.cpp (hunk 5 above)
Comment (NEDJIMAbelgacem) [reply src/3d/qgsshadowrenderingframegraph.cpp:171]: Frame rate does actually increase when I disable the render capture subtree. I added a way to toggle that part on/off.

**Comment by nyalldawson** [comment]
@NEDJIMAbelgacem with your fixes, is it now possible to remove all the

```
  if ( QgsTest::isCIRun() )
  {
    QSKIP( "Intermittently fails on CI" );
  }
```

test skips in testqgs3drenderering.cpp ?

**Comment by NEDJIMAbelgacem** [comment]
> @NEDJIMAbelgacem with your fixes, is it now possible to remove all the
> 
> ```
>   if ( QgsTest::isCIRun() )
>   {
>     QSKIP( "Intermittently fails on CI" );
>   }
> ```
> 
> test skips in testqgs3drenderering.cpp ?

I removed that and tests are passing :)

File: src/app/3d/qgs3dmapcanvas.cpp [hunk 6]
```diff
@@ -177,9 +172,15 @@ void Qgs3DMapCanvas::saveAsImage( const QString fileName, const QString fileForm
...
-    mEngine->requestCaptureImage();
+    QgsOffscreen3DEngine engine;
+    engine.setSize( size() );
+    Qgs3DMapScene scene( *mMap, &engine );
+    engine.setRootEntity( &scene );
+    QgsCameraController *cameraController = mScene->cameraController();
+    scene.cameraController()->setLookingAtPoint( cameraController->lookingAtPoint(), cameraController->distance(), cameraController->pitch(), cameraController->yaw() );
+    QImage img = Qgs3DUtils::captureSceneImage( engine, &scene );
+    img.save( fileName, fileFormat.toLocal8Bit().data() );
+    emit savedAsImage( fileName );
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmapcanvas.cpp:183]: Hmm, this means that a copy of the whole scene needs to be created, and possibly many chunks need to be loaded - that could possibly take many seconds, and consume (temporarily) a lot of GPU memory. Are we sure we want to do it? Why not just use the current 3D engine and capture the image?

File: src/3d/qgsshadowrenderingframegraph.cpp [hunk 7]
```diff
@@ -406,4 +446,18 @@ void QgsShadowRenderingFrameGraph::setSize( QSize s )
...
+
+void QgsShadowRenderingFrameGraph::setRenderCaptureEnabled( bool enabled )
+{
+  if ( enabled == mRenderCaptureEnabled )
+    return;
+  mRenderCaptureEnabled = enabled;
+  if ( mRenderCaptureEnabled )
+    mRenderCaptureTargetSelector->setParent( mPostprocessPassLayerFilter );
+  else
+    mRenderCaptureTargetSelector->setParent( ( Qt3DCore::QNode * ) nullptr );
```
Comment (wonder-sk) [inline src/3d/qgsshadowrenderingframegraph.cpp:462]: Can't we simply call `mRenderCaptureTargetSelector->setEnabled( mRenderCaptureEnabled )` to get the expected result instead of removing parent and then re-assigning it back?

File: src/3d/qgsoffscreen3dengine.h [hunk 8]
```diff
@@ -100,21 +97,7 @@ class _3D_EXPORT QgsOffscreen3DEngine : public QgsAbstract3DEngine
...
-
-    // frame graph stuff
-    Qt3DRender::QRenderSurfaceSelector *mSurfaceSelector = nullptr;
-    Qt3DRender::QRenderTargetSelector *mRenderTargetSelector = nullptr;
-    Qt3DRender::QViewport *mViewport = nullptr;
-    Qt3DRender::QClearBuffers *mClearBuffers = nullptr;
-    Qt3DRender::QCameraSelector *mCameraSelector = nullptr;
-    Qt3DRender::QRenderCapture *mRenderCapture = nullptr;          // The render capture node, which is appended to the frame graph.
-
+//    QgsShadowRenderingFrameGraph *mFrameGraph = nullptr;
```
Comment (wonder-sk) [inline src/3d/qgsoffscreen3dengine.h:100]: commented out code :-)

**Review by wonder-sk — APPROVED** [review]
